contributors/design-proposals/auditing.md

```diff
@@ -0,0 +1,290 @@
+- Talk from IBM: An Introduction to DMTF Cloud Auditing using
+the CADF Event Model and Taxonomies https://wiki.openstack.org/w/images/e/e1/Introduction_to_Cloud_Auditing_using_CADF_Event_Model_and_Taxonomy_2013-10-22.pdf
+
+## Architecture
+
+When implementing audit logging there are basically two options:
+
+1. put a logging proxy in front of the apiserver
+2. integrate audit logging into the apiserver itself
+
+Both approaches have advantages and disadvanteges:
+
+- **pro proxy**:
+  + keeps complexity out of the apiserver
+  + reuses existing solutions
+- **contra proxy**:
+  + has no deeper insight into the Kubernetes api
+  + has no knowledge of auth, authn, admission
+  + has no access to the storage level for differential output
+  + has to terminate SSL and complicates client certificates based auth
+
+In the following the second approach is described without a proxy.
+
+## Proposed Design
+
+The main concepts are those of
+
+- an audit *event*,
+- an audit *level*,
+- an audit *filters*,
+- an audit *output backend*.
+
+An audit event holds all the data necessary for an *output backend* to produce an audit log entry. The *event* is independent of the *output backend*.
+
+The audit event struct is passed through the apiserver layers as an `*audit.Event` pointer inside the apiserver's `Context` object. It is `nil` when auditing is disabled.
+
+If auditing is enabled, the http handler will attach an `audit.Event` to the context:
+
+```go
+func WithAuditEvent(parent Context, e *audit.Event) Context
+func AuditEventFrom(ctx Context) (*audit.Event, bool)
+```
+
+Depending on the audit level (see below), different layers of the apiserver (e.g. http handler, storage) will fill the `audit.Event` struct. Certain fields might stay empty or `nil` if given level does not require that field. E.g. in the case when only http headers are supposed to be audit logged, no `OldObject` or `NewObject` is to be retrieved on the storage layer.
+
+Proposed audit levels are:
+
+- `HttpHeader` - reflects the current level of auditing, iow. provides following information about each request: timestamp, source IP, HTTP method, user info (including group as user and as group), namespace, URI and response code.
+- `RequestObject` - additionally provides unstructured object that was part of the request.
+- `StorageObject` - provides the object before and after modification or the difference of the changes (to be decided).
+
+In addition to the amount of details expressed as an audit level, audit filters allows limiting the scope of auditing operations more granularly.
+
+Proposed filters include:
+- user - logs information coming only from specified user(s).
+- namespace - logs information targeting only specified namespace(s).
+- http method - logs information performed only by specified http method(s).
+- resource - logs information targeting specified resource type.
+
+Specified filters will be AND-ed together when processing requests.
```

Comment:
What if I wanted to specify something like "drop all get requests except get secrets"?